src/runtime/dsl.ts

```diff
@@ -0,0 +1,431 @@
+    let bIsRegister = isRegister(bValue);
+    if(aIsRegister && bIsRegister) {
+      let aVars = this.variableLookup[a.id];
+      let bVars = this.variableLookup[b.id];
+      for(let variable of aVars) {
+        variable.value = bValue;
+        bVars.push(variable);
+      }
+      this.variableLookup[a.id] = [];
+    } else if(aIsRegister) {
+      let aVars = this.variableLookup[a.id];
+      for(let variable of aVars) {
+        variable.value = bValue;
+      }
+      this.variableLookup[a.id] = [];
+    } else if(bIsRegister) {
+      let bVars = this.variableLookup[b.id];
+      for(let variable of bVars) {
+        variable.value = aValue;
+      }
+      this.variableLookup[b.id] = [];
+    } else if(aValue !== bValue) {
+      throw new Error(`Trying to equivalence two static values that aren't the same: ${aValue} and ${bValue}`);
+    }
+  }
+
+  prepare() {
+    let functions = this.functions.slice() as (DSLFunction | undefined)[];
+    // look through the functions for equivalences before we do anything else.
+    let ix = 0;
+    for(let func of functions) {
+      if(!func || func.path[1] !== "==") continue;
+      this.equivalence(func.args[0], func.args[1]);
+      functions[ix] = undefined;
+      ix++;
+    }
+    let registerIx = 0;
+    for(let id in this.variableLookup) {
+      let variable = this.variableLookup[id][0] as DSLVariable;
+      if(!variable || !isRegister(variable.value)) continue;
+      if(variable.value.offset !== -1) throw new Error("We've somehow already assigned a variable's register");
+      variable.value.offset = registerIx++;
+    }
+    let items = functions.concat(this.records as any[]);
+    let constraints = [];
+    let nodes = [];
+    for(let toCompile of items) {
+      if(!toCompile) continue;
+      let compiled = toCompile.compile();
+      if(!compiled) continue;
+      for(let item of compiled) {
+        if(item instanceof Scan || item instanceof FunctionConstraint) {
+          constraints.push(item);
+        } else {
+          nodes.push(item as Node);
+        }
+      }
+    }
+    nodes.unshift(new runtime.JoinNode(constraints))
+    this.block = new runtime.Block(this.name, nodes);
+    console.log(this);
+  }
+
+  //-------------------------------------------------------------------
+  // Code transforms
+  //--------------------------------------------------------------------
+
+  transformBlockCode(code:string, functionArgs:string[]):string {
+
+    let libArg = functionArgs[2];
+    let hasChanged = true;
+    let infixParam = "((?:(?:[a-z0-9_\.]+(?:\\[\".*?\"\\])?)+(?:\\(.*\\))?)|\\(.*\\))";
+    let stringPlaceholder = "(____[0-9]+____)";
+
+    let strings:string[] = [];
+    code = code.replace(/"(?:[^"\\]|\\.)*"/gi, function(str) {
+      strings.push(str);
+      return "____" + (strings.length - 1) + "____";
+    })
+
+    let stringAddition = new RegExp(`(?:${infixParam}\\s*\\+\\s*${stringPlaceholder})|(?:${stringPlaceholder}\\s*\\+\\s*${infixParam})`,"gi");
```

Comment:
Let's show representative examples of the patterns we expect to catch here, like we do for the infixes below.